### DISCUSSION TITLE
Micropython camera API

### DISCUSSION THREAD

**Discussion by cnadler86** [micropython/micropython#15985] [Show and tell]
This is not the first implementation of a camera driver for Micropython, but it is the first aiming to be a general one for various Ports. I used circuitpythons implementation for the esp32 as basis. It is not perfect, but it is IMHO a good starting point. I have precompiled binaries for the generic esp32 boards, but all other boards with a built-in camera Interface will follow. 

[Here the link](https://github.com/cnadler86/micropython-camera-API) for those interested. Have fun 🎢

**Comment by jonnor** [comment]
I tested it with an OV2640 and it works nicely! Hope to build a project with it soon :) Thank you @cnadler86

**Reply by davefes** [reply]
@jonnor, which bit of hardware did you test this on?  A ESP32-CAM AI Thinker?

**Reply by jonnor** [reply]
I used the LilyGo T-Camera Mic v1.6 (quite old, no longer for sale). Have also ordered a LilyGo T-Camera S3 and will test when that arrives :)

**Comment by jonnor** [comment]
I tested on LilyGo TCamera Mic v 1.6. That one is obsolete now, so I have
ordered LilyGo TCamera SE and will test that when it arrives in a few weeks.

**Reply by davefes** [reply]
Thanks.

If somebody does it for the ESP32 AI Thinker (probably clones) I'd would like see how they set up the pins.

**Reply by cnadler86** [reply]
This is a work in progress, but [here](https://github.com/cnadler86/micropython-camera-API/actions/runs/11320499361#artifacts) are precompiled images (will expire in a few days) for many boards with hopefully the right pin assignment, also the AI thinker. It uses the respective generic board as basis. So, if you want, you can test it on your Board. 
I only tested the image for my xiao-board, but why should not work what worked in one sample 🤪. 

Using these makes it easier to construct the camera, since you don't need to assign the pins any more, and you don't need to build the code to predefined the pins. Sadly Micropython doesn't include as many boards as circuitpython does,  but the generic configuration should fit in most if the cases.

**Reply by sophiedegran** [reply]
@davefes look at the take_picture.c in github.com/espressif/esp-camera/blob/master/examples/camera_example/main/ line 74 their define camera pin for BOARD_ESP32CAM_AITHINKER
Hope that help you

**Reply by davefes** [reply]
@cnadler86 thank you for the firmware image I will test it later today.
@sophiedegran thank you.
https://github.com/espressif/esp32-camera/blob/master/examples/camera_example/main/take_picture.c

**Reply by davefes** [reply]
Following the above link it looks like I have a start:
```py
from camera import Camera, GrabMode, PixelFormat, FrameSize, GainCeiling

# Camera construction and initialization
camera = Camera(
    data_pins=[5,18,19,21,36,39,34,35],
    vsync_pin=25,
    href_pin=23,
    sda_pin=26,
    scl_pin=27,
    pclk_pin=22,
    xclk_pin=0,
    xclk_freq=20000000,
    powerdown_pin=32,
    reset_pin=-1,
    pixel_format=PixelFormat.JPEG,
    frame_size=FrameSize.UXGA,
    jpeg_quality=5,
    fb_count=1,
    grab_mode=GrabMode.WHEN_EMPTY
)

img = camera.capture()
```
I end up with a 43K JPEG image but it looks quite pixelated for that size.  I usually grab 3 frames for a snapshot and save the last so will try that.

Changing `jpeg_quality` doesn't seem to make any difference.  I usually get a 150K image with the other camera distro.

Thanks again @sophiedegran for the link.

**Reply by davefes** [reply]
Is my poor quality related to not handling PSRAM properly:
`fb_location = CAMERA_FB_IN_PSRAM,`

**Reply by cnadler86** [reply]
@davefes Hi,
Your poor quality can be improved very easy, just adapt it to jpeg_quality=85. I forgot to tell you to look in the develop branch for the documentation, sorry. I changes the quality definition to the standard used from 0-100%.
With the precompiled FW you should be able to construct the camera without any pin assignment, e.g.:
```
from camera import Camera, GrabMode, PixelFormat, FrameSize, GainCeiling

# Camera construction and initialization
camera = Camera(pixel_format=PixelFormat.JPEG, frame_size=FrameSize.UXGA, jpeg_quality=85)
img = camera.capture()
```
Let me know if this helped out and if you can construct without pin assignment.

**Reply by davefes** [reply]
@cnadler86 Much better.  The other distro had highest quality as a low number. 85 gives a file size of 130K which is roughly what I had before.  Yes, your constructor works.

I am now getting a green tint with the following capture:
```py
          # throw-away two images, this sets
          # the camera up for a proper picture
            unwanted_img = camera.capture()

            time.sleep_ms(100)

            unwanted_img = camera.capture()

            time.sleep_ms(100)

            img = camera.capture()

            print('picture taken')
            time.sleep_ms(10)
```
which worked for me on the other distro, I will go back to taking 5 or 10 frames before saving the last one.

**Reply by cnadler86** [reply]
Try first changing the fb_count=2 and/or grab_mode=GrabMode.LATEST. I am not sure if the esp32 gets to hot or hangs by this configuration, but it might help.

**Reply by cnadler86** [reply]
You could also try another frequency, the default for the ESP32 is 10MHz at the moment

**Reply by davefes** [reply]
Tried `fb_count=2`, `grab_mode=GrabMode.LATEST` and `xclk_freq=20000000` and consistently end up with a pink-tinged picture.  Taking 5 frames didn't help.

Let me try another few cameras.

**Reply by cnadler86** [reply]
I think this is more a camera settings thing. Here is what I found in stack overflow:

"To remove the green tint try disabling all of the auto functions except lenc, and set the sensor settings after initialisation to white balance mode 3 or 4, it should get rid of the green tint. The sensor initialises with AGC and AWB enabled so you need to disable them and fix the white balance before taking a picture."

For this, use the respective set_xxx method.

If you still experience problems with the API, please open an issue there.

**Reply by davefes** [reply]
This is the thread I found:
https://github.com/espressif/esp32-camera/issues/314
Will move to Issues.

Thank you.

**Comment by jkorte-dev** [comment]
Really interesting project! Do you have any plans to port the driver to the stm32h7xxx ?

**Reply by cnadler86** [reply]
Hi,
at the moment, I personally don't. I looked very quickly at [OpenMV](https://github.com/openmv) (since they use micropython and STM32), and I also made a quick github search, but I sadly didn't found any driver equivalent to the esp32-camera driver. Porting the API to the STM32 is therefore associated with more effort and exceeds my current capacity. Perhaps we just need to wait until a similar driver is available (if someone knows one, pleas let us know). But, 6 months back I didn't have any plans in starting a MP Camera API....

And to be honest, I hope someone with more experience with the STM32 (I have zero) contributes this feature. Open source projects live from the community and I am convinced that teamwork and distributing the work across several shoulders generates a better outcome compared to a one-man-show. Time will tell...
